File: benchmarks/python/benchmark_triple_diff.py
#!/usr/bin/env python3
"""
Benchmark: TripleDifference (diff-diff TripleDifference class).

Usage:
    python benchmark_triple_diff.py --data path/to/data.csv --output path/to/results.json \
        [--method dr|reg|ipw] [--covariates]
"""

import argparse
import json
import os
import sys
from pathlib import Path

# IMPORTANT: Parse --backend and set environment variable BEFORE importing diff_diff
def _get_backend_from_args():
    """Parse --backend argument without importing diff_diff."""
    parser = argparse.ArgumentParser(add_help=False)
    parser.add_argument("--backend", default="auto", choices=["auto", "python", "rust"])
    args, _ = parser.parse_known_args()
    return args.backend

_requested_backend = _get_backend_from_args()
if _requested_backend in ("python", "rust"):
    os.environ["DIFF_DIFF_BACKEND"] = _requested_backend

# NOW import diff_diff and other dependencies (will see the env var)
import pandas as pd

# Add parent to path for imports
sys.path.insert(0, str(Path(__file__).parent.parent.parent))

from diff_diff import TripleDifference
from benchmarks.python.utils import Timer


def parse_args():
    parser = argparse.ArgumentParser(description="Benchmark TripleDifference estimator")
    parser.add_argument("--data", required=True, help="Path to input CSV data")
    parser.add_argument("--output", required=True, help="Path to output JSON results")
    parser.add_argument(
        "--method", default="dr", choices=["dr", "reg", "ipw"],
        help="Estimation method: dr (default), reg, ipw"
    )
    parser.add_argument(
        "--covariates", action="store_true",
        help="Include covariates (columns starting with 'cov')"
    )
    parser.add_argument(
        "--backend", default="auto", choices=["auto", "python", "rust"],
        help="Backend to use: auto (default), python (pure Python), rust (Rust backend)"
    )
    return parser.parse_args()


def main():
    args = parse_args()

    print(f"Loading data from: {args.data}")
    data = pd.read_csv(args.data)

    # Map R column names to Python convention
    column_map = {"y": "outcome", "state": "group", "id": "unit_id"}
    data = data.rename(columns=column_map)

    # Map R time encoding {1, 2} to Python {0, 1}
    if data["time"].min() == 1:
        data["time"] = data["time"] - 1

    # Detect covariate columns
    cov_cols = [c for c in data.columns if c.startswith("cov")]
    covariates = cov_cols if args.covariates and cov_cols else None

    print(f"Running DDD estimation (method={args.method}, "
          f"covariates={covariates is not None})...")

    ddd = TripleDifference(estimation_method=args.method)

    with Timer() as timer:
        results = ddd.fit(
            data,
            outcome="outcome",
            group="group",
            partition="partition",
            time="time",
            covariates=covariates,
        )

    # Compute CI bounds
    ci_lower, ci_upper = results.conf_int

    output = {
        "ATT": results.att,
        "se": results.se,
        "lci": ci_lower,
        "uci": ci_upper,
        "method": args.method,
        "covariates": args.covariates,
        "n_obs": results.n_obs,
        "elapsed_seconds": timer.elapsed,
    }

    print(f"Writing results to: {args.output}")
    with open(args.output, "w") as f:
        json.dump(output, f, indent=2)

    print("Done.")
    print(f"  ATT = {results.att:.6f}")
    print(f"  SE  = {results.se:.6f}")
    print(f"  Time: {timer.elapsed:.3f}s")


if __name__ == "__main__":
    main()

File: benchmarks/R/benchmark_triplediff.R
#!/usr/bin/env Rscript
# Benchmark: Triple Difference (R `triplediff` package)
#
# This uses triplediff::ddd() with panel=FALSE (repeated cross-section mode),
# matching the Python TripleDifference estimator's approach.
#
# Usage:
#   Rscript benchmark_triplediff.R --data path/to/data.csv --output path/to/results.json \
#     [--method dr|reg|ipw] [--covariates true|false]

library(triplediff)
library(jsonlite)
library(data.table)

# Parse command line arguments
args <- commandArgs(trailingOnly = TRUE)

parse_args <- function(args) {
  result <- list(
    data = NULL,
    output = NULL,
    method = "dr",
    covariates = FALSE
  )

  i <- 1
  while (i <= length(args)) {
    if (args[i] == "--data") {
      result$data <- args[i + 1]
      i <- i + 2
    } else if (args[i] == "--output") {
      result$output <- args[i + 1]
      i <- i + 2
    } else if (args[i] == "--method") {
      result$method <- args[i + 1]
      i <- i + 2
    } else if (args[i] == "--covariates") {
      result$covariates <- tolower(args[i + 1]) == "true"
      i <- i + 2
    } else {
      i <- i + 1
    }
  }

  if (is.null(result$data) || is.null(result$output)) {
    stop("Usage: Rscript benchmark_triplediff.R --data <path> --output <path> [--method dr|reg|ipw] [--covariates true|false]")
  }

  return(result)
}

config <- parse_args(args)

# Load data
message(sprintf("Loading data from: %s", config$data))
data <- fread(config$data)

# Build covariate formula
cov_cols <- grep("^cov", names(data), value = TRUE)
if (config$covariates && length(cov_cols) > 0) {
  xformla <- as.formula(paste("~", paste(cov_cols, collapse = "+")))
  message(sprintf("Using covariates: %s", paste(cov_cols, collapse = ", ")))
} else {
  xformla <- ~1
  message("No covariates")
}

# Run benchmark
message(sprintf("Running DDD estimation (method=%s, panel=FALSE)...", config$method))
timing <- system.time({
  res <- ddd(
    yname = "y",
    tname = "time",
    idname = "id",
    gname = "state",
    pname = "partition",
    data = data,
    control_group = "nevertreated",
    panel = FALSE,
    xformla = xformla,
    est_method = config$method,
    boot = FALSE
  )
})

# Collect results
output <- list(
  ATT = res$ATT,
  se = res$se,
  lci = res$lci,
  uci = res$uci,
  method = config$method,
  covariates = config$covariates,
  n_obs = nrow(data),
  elapsed_seconds = timing["elapsed"]
)

# Write results
message(sprintf("Writing results to: %s", config$output))
write(toJSON(output, pretty = TRUE, auto_unbox = TRUE, digits = 15), config$output)

message("Done.")
message(sprintf("  ATT = %.6f", res$ATT))
message(sprintf("  SE  = %.6f", res$se))
message(sprintf("  Time: %.3fs", timing["elapsed"]))
